pkg/applyconfigurations/gen.go

```diff
@@ -0,0 +1,238 @@
+}
+
+// enableOnType marks whether applyconfiguration generation is enabled for the type.
+func enabledOnType(info *markers.TypeInfo) bool {
+	if typeMarker := info.Markers.Get(enableTypeMarker.Name); typeMarker != nil {
+		return typeMarker.(bool)
+	}
+	return isCRD(info)
+}
+
+func outputPkg(col *markers.Collector, pkg *loader.Package) (string, error) {
+	pkgMarkers, err := markers.PackageMarkers(col, pkg)
+	if err != nil {
+		return "", err
+	}
+	pkgMarker := pkgMarkers.Get(outputPkgMarker.Name)
+	if pkgMarker != nil {
+		return pkgMarker.(string), nil
+	}
+	return "", nil
+}
+
+// isCRD marks whether the type is a CRD based on the +kubebuilder:resource marker.
+func isCRD(info *markers.TypeInfo) bool {
+	objectEnabled := info.Markers.Get(isCRDMarker.Name)
+	return objectEnabled != nil
+}
+
+func (d Generator) Generate(ctx *genall.GenerationContext) error {
+	headerFilePath := d.HeaderFile
+
+	if headerFilePath == "" {
+		tmpFile, err := os.CreateTemp("", "applyconfig-header-*.txt")
+		if err != nil {
+			return fmt.Errorf("failed to create temporary file: %w", err)
+		}
+		tmpFile.Close()
+
+		defer os.Remove(tmpFile.Name())
+
+		headerFilePath = tmpFile.Name()
+	}
+
+	objGenCtx := ObjectGenCtx{
+		Collector:      ctx.Collector,
+		Checker:        ctx.Checker,
+		HeaderFilePath: headerFilePath,
+	}
+
+	for _, pkg := range ctx.Roots {
+		if err := objGenCtx.generateForPackage(pkg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ObjectGenCtx contains the common info for generating apply configuration implementations.
+// It mostly exists so that generating for a package can be easily tested without
+// requiring a full set of output rules, etc.
+type ObjectGenCtx struct {
+	Collector      *markers.Collector
+	Checker        *loader.TypeChecker
+	HeaderFilePath string
+}
+
+// generateForPackage generates apply configuration implementations for
+// types in the given package, writing the formatted result to given writer.
+// May return nil if source could not be generated.
+func (ctx *ObjectGenCtx) generateForPackage(root *loader.Package) error {
+	enabled, _ := enabledOnPackage(ctx.Collector, root)
+	if !enabled {
+		return nil
+	}
+	if len(root.GoFiles) == 0 {
+		return nil
+	}
+
+	genericArgs, _ := generatorargs.NewDefaults()
+	genericArgs.InputDirs = []string{root.PkgPath}
+	genericArgs.GoHeaderFilePath = ctx.HeaderFilePath
+
+	outpkg, _ := outputPkg(ctx.Collector, root)
+	if outpkg == "" {
+		outpkg = importPathSuffix
+	}
+	genericArgs.OutputPackagePath = filepath.Join(root.PkgPath, outpkg)
+
+	// attempts to retrieve the correct base directory to output apply configurations to by
+	// looking into the package and retrieving the first go file it finds, and using that as the output base.
+	// this is because we cannot rely on gogen calculating the correct output base.
+	// if we leave this empty, gogen will attempt to use GOPATH to write the files which is not wanted
+	genericArgs.OutputBase = filepath.Dir(root.GoFiles[0])
```

Comment:
I think with some more invest it may be possible to resolve the directory smoother but there we issues with the output package not having correct import paths. For now this will work fine with the generation of the base directory being in the package path. For the request of having a directory outside of the package path this needs adjustment though. Open to any ideas here

Reply:
I believe in general it should now work both within and outside GOPATH and should resolve the issue @sbueringer and @alvaroaleman not seeing the generated files in the right folders